substrate/dex/pallet/src/lib.rs

```diff
@@ -161,29 +161,75 @@ pub mod pallet {
   /// Moving window of oracle prices.
   ///
   /// The second [u8; 8] key is the amount's big endian bytes, and u16 is the amount of inclusions
-  /// in this multi-set.
+  /// in this multi-set. Since the underlying map lexiogracpilly sorted, this map stores amounts
+  /// from low to high.
   #[pallet::storage]
-  #[pallet::getter(fn oracle_prices)]
   pub type OraclePrices<T: Config> =
     StorageDoubleMap<_, Identity, Coin, Identity, [u8; 8], u16, OptionQuery>;
-  impl<T: Config> Pallet<T> {
-    // TODO: consider an algorithm which removes outliers? This algorithm might work a good bit
-    // better if we remove the bottom n values (so some value sustained over 90% of blocks instead
-    // of all blocks in the window).
-    /// Get the highest sustained value for this window.
-    /// This is actually the lowest price observed during the windows, as it's the price
-    /// all prices are greater than or equal to.
-    pub fn highest_sustained_price(coin: &Coin) -> Option<Amount> {
-      let mut iter = OraclePrices::<T>::iter_key_prefix(coin);
-      // the first key will be the lowest price due to the keys being lexicographically ordered.
-      iter.next().map(|amount| Amount(u64::from_be_bytes(amount)))
-    }
-  }
+
+  /// Same as `OraclePrices` but in reverse order, since the keys we save to this map are
+  /// lexiogracpilly reversed. Hence this map stores the amounts from hight to low.
+  #[pallet::storage]
+  pub type OraclePricesReverse<T: Config> =
+    StorageDoubleMap<_, Identity, Coin, Identity, [u8; 8], u16, OptionQuery>;
 
   #[pallet::storage]
   #[pallet::getter(fn oracle_value)]
   pub type OracleValue<T: Config> = StorageMap<_, Identity, Coin, Amount, OptionQuery>;
 
+  /// Current median price of the prices in the oracle prices maps at any given time.
+  #[pallet::storage]
+  #[pallet::getter(fn median_price)]
+  pub type MedianPrice<T: Config> = StorageMap<_, Identity, Coin, Amount, OptionQuery>;
+  impl<T: Config> Pallet<T> {
+    /// Update the current median according to new state of oracle prices.
+    pub fn update_median_price(
+      coin: &Coin,
+      inserted: u64,
+      removed: Option<u64>,
+      n: BlockNumberFor<T>,
+    ) {
+      let current = Self::median_price(coin).unwrap_or(Amount(0)).0;
+
+      // if we didn't remove  an amount from the window, just means we still didn't hit
+      // the full size of the window.
+      let next = if removed.is_none() {
+        if n % 2u32.into() == 1u32.into() {
+          if inserted >= current {
+            OraclePrices::<T>::iter_key_prefix_from(coin, current.to_be_bytes().to_vec()).next()
+          } else {
+            let reversed = reverse_lexicographic_order(current.to_be_bytes());
+            OraclePricesReverse::<T>::iter_key_prefix_from(coin, reversed.to_vec())
+              .next()
+              .map(reverse_lexicographic_order)
+          }
+        } else {
+          Some(current.to_be_bytes())
```

Comment:
Let's document that if there's an even amount of items, there's two candidates and the current one remains a valid candidate.